### PR TITLE
CMake Build System Support for SoftHSM

### PR DESCRIPTION
**Pull request by conz27**
Introducing a more modern & parallel build system for compiling SoftHSM.

Key Features:
* finds libraries and header files for OpenSSL, Botan, SQLite3, and CppUnit
* compiles all libraries
* compiles all tests
* integrates tests into CTest for easier execution
* supports most m4/acx_*.m4 configuration checks
* supports install/package/package_source

Missing Features:
* does not support win32

Build Instructions:

1. cmake -H. -Bbuild -DBUILD_TESTS=ON
2. make -C build

Running Tests:

1. cd build/
2. ctest -V

Install:
make -C build install

Package Binaries:
make -C build package

Package Source Code:
make -C build package_source

Other Makefile Targets:
make -C build help

### CODE REVIEW

**Comment by bellgrim** [comment]
Thank you for the PR. Will have a look at this for SoftHSM 2.5.

**Comment by conz27** [comment]
Sounds great @bellgrim . The build system will not be 100% one-to-one with the AutoTools one, but I'll try get it as close as possible.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,25 +1,59 @@
...
-
 # Build Options
 option(BUILD_TESTS "Compile tests along with libraries" OFF)
-option(DISABLE_NON_PAGED_MEMORY "Disable non-paged memory for secure storage" ON)
-option(DISABLE_VISIBILITY "Disables and unsets -fvisibility=hidden")
+option(DISABLE_NON_PAGED_MEMORY "Disable non-paged memory for secure storage" OFF)
+option(DISABLE_VISIBILITY "Disables and unsets -fvisibility=hidden" OFF)
 option(ENABLE_64bit "Enable 64-bit compiling" OFF)
 option(ENABLE_ECC "Enable support for ECC" ON)
 option(ENABLE_GOST "Enable support for GOST" ON)
```
Comment (dmercer-google) [inline CMakeLists.txt]: Shouldn't this be off by default as GOST is not a standard part of OpenSSL?

File: modules/CompilerOptions.cmake [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    if(CPPUNIT_FOUND)
+        message(STATUS "Found CppUnit")
+        set(CPPUNIT_INCLUDES ${CPPUNIT_INCLUDE_DIR})
+        set(CPPUNIT_LIBS ${CPPUNIT_LIBRARY})
+    else(CPPUNIT_FOUND)
+        message(FATAL_ERROR "Failed to find CppUnit!")
+    endif(CPPUNIT_FOUND)
+endif(BUILD_TESTS)
+
+configure_file(config.h.in.cmake ${CMAKE_BINARY_DIR}/config.h)
```
Comment (dmercer-google) [inline modules/CompilerOptions.cmake]: This file does not seem to exist.  Changing to config.h.in seems to work

Comment (conz27) [reply modules/CompilerOptions.cmake]: git for some reason didn't detect it, so I had to force add it.

File: CMakeLists.txt (hunk 1 above)
Comment (conz27) [reply CMakeLists.txt]: I was using the output of `./configure --help` to determine how things are set by default:

```
./configure --help
`configure' configures SoftHSM 2.3.0 to adapt to many kinds of systems.

Usage: ./configure [OPTION]... [VAR=VALUE]...

To assign environment variables (e.g., CC, CFLAGS...), specify them as
VAR=VALUE.  See below for descriptions of some of the useful variables.

Defaults for the options are specified in brackets.

Configuration:
  -h, --help              display this help and exit
      --help=short        display options specific to this package
      --help=recursive    display the short help of all the included packages
  -V, --version           display version information and exit
  -q, --quiet, --silent   do not print `checking ...' messages
      --cache-file=FILE   cache test results in FILE [disabled]
  -C, --config-cache      alias for `--cache-file=config.cache'
  -n, --no-create         do not create output files
      --srcdir=DIR        find the sources in DIR [configure dir or `..']

Installation directories:
  --prefix=PREFIX         install architecture-independent files in PREFIX
                          [/usr/local]
  --exec-prefix=EPREFIX   install architecture-dependent files in EPREFIX
                          [PREFIX]

By default, `make install' will install all the files in
`/usr/local/bin', `/usr/local/lib' etc.  You can specify
an installation prefix other than `/usr/local' using `--prefix',
for instance `--prefix=$HOME'.

For better control, use the options below.

Fine tuning of the installation directories:
  --bindir=DIR            user executables [EPREFIX/bin]
  --sbindir=DIR           system admin executables [EPREFIX/sbin]
  --libexecdir=DIR        program executables [EPREFIX/libexec]
  --sysconfdir=DIR        read-only single-machine data [PREFIX/etc]
  --sharedstatedir=DIR    modifiable architecture-independent data [PREFIX/com]
  --localstatedir=DIR     modifiable single-machine data [PREFIX/var]
  --runstatedir=DIR       modifiable per-process data [LOCALSTATEDIR/run]
  --libdir=DIR            object code libraries [EPREFIX/lib]
  --includedir=DIR        C header files [PREFIX/include]
  --oldincludedir=DIR     C header files for non-gcc [/usr/include]
  --datarootdir=DIR       read-only arch.-independent data root [PREFIX/share]
  --datadir=DIR           read-only architecture-independent data [DATAROOTDIR]
  --infodir=DIR           info documentation [DATAROOTDIR/info]
  --localedir=DIR         locale-dependent data [DATAROOTDIR/locale]
  --mandir=DIR            man documentation [DATAROOTDIR/man]
  --docdir=DIR            documentation root [DATAROOTDIR/doc/softhsm]
  --htmldir=DIR           html documentation [DOCDIR]
  --dvidir=DIR            dvi documentation [DOCDIR]
  --pdfdir=DIR            pdf documentation [DOCDIR]
  --psdir=DIR             ps documentation [DOCDIR]

Program names:
  --program-prefix=PREFIX            prepend PREFIX to installed program names
  --program-suffix=SUFFIX            append SUFFIX to installed program names
  --program-transform-name=PROGRAM   run sed PROGRAM on installed program names

System types:
  --build=BUILD     configure for building on BUILD [guessed]
  --host=HOST       cross-compile to build programs to run on HOST [BUILD]

Optional Features:
  --disable-option-checking  ignore unrecognized --enable/--with options
  --disable-FEATURE       do not include FEATURE (same as --enable-FEATURE=no)
  --enable-FEATURE[=ARG]  include FEATURE [ARG=yes]
  --enable-silent-rules   less verbose build output (undo: "make V=1")
  --disable-silent-rules  verbose build output (undo: "make V=0")
  --enable-dependency-tracking
                          do not reject slow dependency extractors
  --disable-dependency-tracking
                          speeds up one-time build
  --enable-pedantic       enable pedantic compile mode [enabled]
  --enable-strict         enable strict compile mode [enabled]
  --enable-64bit          enable 64-bit compiling [disabled]
  --enable-ecc            Enable support for ECC (default enabled)
  --enable-gost           Enable support for GOST (default enabled)
  --enable-fips           Enable support for FIPS 140-2 mode (default
                          disabled)
  --disable-non-paged-memory
                          Disable non-paged memory for secure storage (default
                          enabled)
  --disable-visibility    disable hidden visibilty link mode [enabled]
  --enable-p11-kit        Enable p11-kit integration (default enabled)
  --enable-shared[=PKGS]  build shared libraries [default=yes]
  --enable-static[=PKGS]  build static libraries [default=yes]
  --enable-fast-install[=PKGS]
                          optimize for fast installation [default=yes]
  --disable-libtool-lock  avoid locking (might break parallel builds)

Optional Packages:
  --with-PACKAGE[=ARG]    use PACKAGE [ARG=yes]
  --without-PACKAGE       do not use PACKAGE (same as --with-PACKAGE=no)
  --with-crypto-backend   Select crypto backend (openssl|botan)
  --with-openssl=PATH     Specify prefix of path of OpenSSL
  --with-botan=PATH       Specify prefix of path of Botan
  --with-migrate          Build the migration tool. Requires SQLite3.
  --with-objectstore-backend-db
                          Build with object store backend database (SQLite3).
  --with-sqlite3=PATH     Specify prefix of path of SQLite3
  --with-p11-kit=PATH     Specify install path of the p11-kit module, will
                          override path given by pkg-config
  --with-pic[=PKGS]       try to use only PIC/non-PIC objects [default=use
                          both]
  --with-aix-soname=aix|svr4|both
                          shared library versioning (aka "SONAME") variant to
                          provide on AIX, [default=aix].
  --with-gnu-ld           assume the C compiler uses GNU ld [default=no]
  --with-sysroot[=DIR]    Search for dependent libraries within DIR (or the
                          compiler's sysroot if not specified).

Some influential environment variables:
  CC          C compiler command
  CFLAGS      C compiler flags
  LDFLAGS     linker flags, e.g. -L<lib dir> if you have libraries in a
              nonstandard directory <lib dir>
  LIBS        libraries to pass to the linker, e.g. -l<library>
  CPPFLAGS    (Objective) C/C++ preprocessor flags, e.g. -I<include dir> if
              you have headers in a nonstandard directory <include dir>
  CXX         C++ compiler command
  CXXFLAGS    C++ compiler flags
  CPP         C preprocessor
  LT_SYS_LIBRARY_PATH
              User-defined run-time library search path.
  CXXCPP      C++ preprocessor

Use these variables to override the choices made by `configure' or to help
it to find libraries and programs with nonstandard names/locations.

Report bugs to the package provider.
```

**Comment by conz27** [comment]
@bellgrim , I think the AppVeyor build environment is broken due to missing OpenSSL dependencies. It works fine in TravisCI.

**Comment by conz27** [comment]
Just rebased on top `develop`.

File: modules/CompilerOptions.cmake [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+                )
+        if(NOT COMPILE_RESULT)
+            message(WARN "failed to compile: ${testfile}")
+        endif(NOT COMPILE_RESULT)
+
+        if(HAVE_AES_KEY_WRAP EQUAL 0)
+            message(STATUS "OpenSSL: RFC 3394 is supported")
+        else()
+            set(error_msg "OpenSSL: RFC 3394 is not supported!!")
+            message(WARNING ${error_msg})
```
Comment (conz27) [inline modules/CompilerOptions.cmake]: WARNING -> WARN

File: modules/CompilerOptions.cmake [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        # acx_openssl_rfc3349
+        set(testfile ${CMAKE_SOURCE_DIR}/modules/tests/test_openssl_rfc3394.c)
+        try_run(HAVE_AES_KEY_WRAP COMPILE_RESULT
+                "${CMAKE_BINARY_DIR}/prebuild_santity_tests" ${testfile}
+                LINK_LIBRARIES ${CRYPTO_LIBS}
+                CMAKE_FLAGS
+                    "-DINCLUDE_DIRECTORIES=${CRYPTO_INCLUDES}"
+                )
+        if(NOT COMPILE_RESULT)
+            message(WARN "failed to compile: ${testfile}")
```
Comment (conz27) [inline modules/CompilerOptions.cmake]: WARN -> WARNING

File: modules/CompilerOptions.cmake [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+        # acx_openssl_rfc5649
+        set(testfile ${CMAKE_SOURCE_DIR}/modules/tests/test_openssl_rfc5649.c)
+        try_run(HAVE_AES_KEY_WRAP_PAD COMPILE_RESULT
+                "${CMAKE_BINARY_DIR}/prebuild_santity_tests" ${testfile}
+                LINK_LIBRARIES ${CRYPTO_LIBS}
+                CMAKE_FLAGS
+                    "-DINCLUDE_DIRECTORIES=${CRYPTO_INCLUDES}"
+                )
+        if(NOT COMPILE_RESULT)
+            message(WARN "failed to compile: ${testfile}")
```
Comment (conz27) [inline modules/CompilerOptions.cmake]: WARN -> WARNING

**Comment by bellgrim** [comment]
@conz27 Our AppVeyor CI script will download dependencies from GitHub. PowerShell uses TLS 1.0 by default and GitHub must have disabled that recently. Have fixed this in the develop branch.

**Comment by bellgrim** [comment]
@dmercer-google Work on SoftHSM 2.5 will start within 1-3 weeks.

**Comment by bellgrim** [comment]
I have now started to work on SoftHSM 2.5. CMake is the third item on my backlog for 2.5, so will review and test this once I have finished #368 and #323.

**Pull request merged**